### PR TITLE
Length field delimited framer

### PR DESCRIPTION
**Pull request by carllerche**
Remaining:

- [x] Docs
- [x] Review

### CODE REVIEW

File: src/length_delimited.rs [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+        FramedRead {
+            inner: codec::FramedRead::new(upstream, Decoder {
+                builder: self,
+                state: DecodeState::Head,
+            }),
+        }
+    }
+
+    pub fn writer<T, B>(self, inner: T) -> FramedWrite<T, B>
+        where B: IntoBuf,
```
Comment (alexcrichton) [inline src/length_delimited.rs]: Perhaps `T` could also be bounded here?

File: src/codec.rs [hunk 2]
```diff
@@ -13,3 +13,346 @@
...
+    //!     .new_read(io);
+    //! # }
+    //! ```
+    //!
+    //! The following frame will be decoded as such:
+    //!
+    //! ```text
+    //!          INPUT                        DECODED
+    //! +-- len ---+--- Payload ---+     +--- Payload ---+
+    //! | \x00\x0B |  Hello world  | --> |  Hello world  |
```
Comment (alexcrichton) [inline src/codec.rs]: Could this be swapped with the first example? This one at least seems like the first I'd run into, whereas the first is a bit more "advanced"

**Comment by alexcrichton** [comment]
Oh CI is also failing I believe

**Comment by tikue** [comment]
So it looks like the decoded frames are `BytesMut`? Is it up to the user to decode the length header at that point?

**Comment by alexcrichton** [comment]
@tikue yeah the data is buffered up in a `BytesMut` instance but the header is decoded automatically here. It looks like there's enough parameters though to configure whether the header is returned as part of the `BytesMut` or not manually.

**Pull request merged**